### ISSUE TITLE
Optimize client_ignore_enterleave_events

### ISSUE THREAD

**Issue by psychon** [awesomeWM/awesome#1109]
Currently, `client_ignore_enterleave_events()` goes through all clients and changes the kinds of events that the X server sends to us. As #1107 shows, this can be quite slow (even though I don't completely understand why). The bisect in there pointed at some other commit, but that commit just caused `client_ignore_enterleave_events()` to be called once per main loop iteration.

The above algorithm could be replaced by something more efficient:
We grab the server during calls to these function and remember the sequence numbers of these requests (a NoOp request's sequence number should be used instead of the actual UngrabServer request!). Later in normal event handling we just ignore events whose sequence numbers fall into this range, because they must have been caused by ourselves.
Pseudocode:
```
void client_ignore_enterleave_events(void)
{
  assert(start_sequence == 0);
  start_sequence = xcb_grab_server(globalconf.connection).sequence;
}
void client_restore_enterleave_events(void)
{
  assert(start_sequence != 0);
  end_sequence = xcb_no_operation(globalconf.connection).sequence);
  xcb_ungrab_server(globalconf.connection);
  add_to_list(globalconf.enterleave_ignored, (struct pair) { start_sequence, end_sequence });

  start_sequence = 0;
}
```
(And event handling has to ignore enter/leave events whose sequence number is in the list and it also has to remove old entries from this list).

**Comment by Elv13** [comment]
Hello, you merged a bunch of commits related to this. What's the status of this issue?

**Comment by psychon** [comment]
Whoops. Fixed with 19094de661f05640e570b23c56d4d175d9faaaef.
